hazelcast/src/main/java/com/hazelcast/util/scheduler/SecondsBasedEntryTaskScheduler.java

```diff
@@ -99,303 +89,301 @@ public int compare(ScheduledEntry o1, ScheduledEntry o2) {
+            return 0;
+        }
 
-            if (cancelAndCleanUpIfEmpty(second, entries, compositeKey, entryToRemove)) {
-                cancelled++;
-            }
+        void executed(ScheduledEntry<K, V> entry) {
+            assert entry.getScheduleId() == id;
+            // no need to remove entry from group, whole group is being executed and will be removed
+            keys.remove(key);
         }
-        return cancelled;
     }
 
     /**
-     * Return all composite keys with the given {@code key}
+     * Manages scheduling where each scheduled entry for given key should be executed when its time comes.
+     * @see ScheduleType#FOR_EACH
      */
-    private Set<CompositeKey> getCompositeKeys(K key) {
-        Set<CompositeKey> candidateKeys = new HashSet<CompositeKey>();
-        for (Object keyObj : secondsOfKeys.keySet()) {
-            CompositeKey compositeKey = (CompositeKey) keyObj;
-            if (compositeKey.getKey().equals(key)) {
-                candidateKeys.add(compositeKey);
+    private final class PerKeyForEachScheduler extends PerKeyScheduler {
+        final K key;
+        final Map<Long, ScheduledGroup> idToGroupMap = new HashMap<>();
+
+        PerKeyForEachScheduler(K key) {
+            this.key = key;
+        }
+
+        boolean schedule(ScheduledEntry<K, V> entry, ScheduledGroup group) {
+            Long id = entry.getScheduleId();
+            idToGroupMap.put(id, group);
+            group.addEntry(id, entry);
+            return true;
+        }
+
+        ScheduledEntry<K, V> get() {
+            ScheduledEntry<K, V> entry = null;
+            for (Map.Entry<Long, ScheduledGroup> idToGroup : idToGroupMap.entrySet()) {
+                Long id = idToGroup.getKey();
+                ScheduledGroup group = idToGroup.getValue();
+                entry = group.getEntry(id);
             }
+            assert entry != null;
+            return entry;
         }
-        return candidateKeys;
-    }
 
-    /**
-     * Returns one scheduled entry for the given {@code key} with no guaranteed ordering
-     */
-    public ScheduledEntry<K, V> getByCompositeKey(K key) {
-        Set<CompositeKey> candidateKeys = getCompositeKeys(key);
-        ScheduledEntry<K, V> result = null;
-        for (CompositeKey compositeKey : candidateKeys) {
-            Integer second = secondsOfKeys.get(compositeKey);
-            if (second != null) {
-                Map<Object, ScheduledEntry<K, V>> entries = scheduledEntries.get(second);
-                if (entries != null) {
-                    result = entries.get(compositeKey);
+        ScheduledEntry<K, V> cancel() {
+            ScheduledEntry<K, V> entry = null;
+            for (Map.Entry<Long, ScheduledGroup> idToGroup : idToGroupMap.entrySet()) {
+                Long id = idToGroup.getKey();
+                ScheduledGroup group = idToGroup.getValue();
+                entry = group.removeEntry(id);
+            }
+            keys.remove(key);
```

Comment:
Should we clear the `idToGroup` here as well?

Reply:
It would be superfluous - the line `keys.remove(key)` removes whole object `PerKeyForEachScheduler` which is the owner of `idToGroup` map, so there will be no reference to this object and it will be GC'ed together with the map. I wanted to avoid any unnecessary work in `cancel()` method, as it is on the hot path when it comes to performance.

Reply:
You're right, missed that, thanks!